tensorboard/data/server/server.rs

```diff
@@ -247,26 +247,208 @@ impl TensorBoardDataProvider for DataProviderHandler {
+                }
+                let (mut max_step, mut max_wall_time, mut max_length) = (None, None, None);
+                for (step, wall_time, value) in ts.valid_values() {
+                    if max_step.map_or(true, |s| s < step) {
+                        max_step = Some(step);
+                    }
+                    if max_wall_time.map_or(true, |wt| wt < wall_time) {
+                        max_wall_time = Some(wall_time);
+                    }
+                    if max_length.map_or(true, |len| len < value.0.len()) {
+                        max_length = Some(value.0.len());
+                    }
+                }
+                let (max_step, max_wall_time, max_length) =
+                    match (max_step, max_wall_time, max_length) {
+                        (Some(s), Some(wt), Some(len)) => (s, wt, len),
+                        _ => continue,
+                    };
+                run_res
+                    .tags
+                    .push(data::list_blob_sequences_response::TagEntry {
+                        tag_name: tag.0.clone(),
+                        metadata: Some(data::BlobSequenceMetadata {
+                            max_step: max_step.into(),
+                            max_wall_time: max_wall_time.into(),
+                            max_length: max_length as i64,
+                            summary_metadata: Some(*ts.metadata.clone()),
+                        }),
+                    });
+            }
+            if !run_res.tags.is_empty() {
+                run_res.run_name = run.0.clone();
+                res.runs.push(run_res);
+            }
+        }
+
+        Ok(Response::new(res))
     }
 
     async fn read_blob_sequences(
         &self,
-        _request: Request<data::ReadBlobSequencesRequest>,
+        req: Request<data::ReadBlobSequencesRequest>,
     ) -> Result<Response<data::ReadBlobSequencesResponse>, Status> {
-        Err(Status::unimplemented("not yet implemented"))
+        let req = req.into_inner();
+        let want_plugin = parse_plugin_filter(req.plugin_filter)?;
+        let (run_filter, tag_filter) = parse_rtf(req.run_tag_filter);
+        let num_points = parse_downsample(req.downsample)?;
+        let runs = self.read_runs()?;
+
+        let mut res: data::ReadBlobSequencesResponse = Default::default();
+        for (run, data) in runs.iter() {
+            if !run_filter.want(run) {
+                continue;
+            }
+            let data = data
+                .read()
+                .map_err(|_| Status::internal(format!("failed to read run data for {:?}", run)))?;
+            let mut run_res: data::read_blob_sequences_response::RunEntry = Default::default();
+            for (tag, ts) in &data.blob_sequences {
+                if !tag_filter.want(tag) {
+                    continue;
+                }
+                if plugin_name(&ts.metadata) != Some(&want_plugin) {
+                    continue;
+                }
```

Comment:
Yeah; I’ve been playing around with this in my head but haven’t actually
sketched it out. I agree both that repetition isn’t intrinsically bad
and that it feels like this common behavior could be factored out to
save a bunch of lines.  It’s made a bit tricky because the response
types (like `data::read_blob_sequences_response::RunEntry`) all differ,
and RPCs access different commit fields (e.g., `&data.blob_sequences`).

I also agree that some kind of iterable would make sense: something like

```rust
for item in self.iter_data(rtf, plugin_filter, |rd| &rd.blob_sequences) {
    let (run, tag, time_series) = item?; // might return "failed to read..."
    // ...
}
```

with the catch that inside the loop we wouldn’t have `run_res` like we
do right now.

I was planning to wait until we have the three data classes implemented
before trying to pull things out. What do you think? Any suggestions?

Reply:
No immediate suggestion. I will let you decide what should be done at places where we do `read_runs` and use run_tag_filter.